ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,864 @@
+        // using this as the end offset for delete includes the whole index segment in the search.
+        cutoffOffset = indexSegment.getStartOffset();
+        break;
+      }
+    }
+    return cutoffOffset;
+  }
+
+  /**
+   * Copies data from the provided log segment into the target log (swap spaces).
+   * @param logSegmentToCopy the {@link LogSegment} to copy from.
+   * @param duplicateSearchSpan the {@link FileSpan} in which to search for duplicates.
+   * @param startOffsetOfLastIndexSegmentForDelete  the start {@link Offset} of the index segment up until which delete
+   *                                                records are considered applicable.
+   * @return {@code true} if all the records in the log segment were copied. {@code false} if some records were not
+   * copied either because there was no more capacity or because a shutdown was initiated.
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there are any problems reading or writing to store components.
+   */
+  private boolean copyLogSegmentData(LogSegment logSegmentToCopy, FileSpan duplicateSearchSpan,
+      Offset startOffsetOfLastIndexSegmentForDelete) throws IOException, StoreException {
+    boolean allCopied = true;
+    for (Offset indexSegmentStartOffset : getIndexSegmentDetails(logSegmentToCopy.getName()).keySet()) {
+      IndexSegment indexSegmentToCopy = srcIndex.getIndexSegments().get(indexSegmentStartOffset);
+      if (needsCopying(indexSegmentToCopy.getEndOffset()) && !copyIndexSegmentData(logSegmentToCopy, indexSegmentToCopy,
+          duplicateSearchSpan, startOffsetOfLastIndexSegmentForDelete)) {
+        // there is a shutdown in progress or there was no space to copy all entries.
+        allCopied = false;
+        break;
+      }
+    }
+    return allCopied;
+  }
+
+  /**
+   * Copies data in the provided {@code indexSegmentToCopy} into the target log (swap spaces).
+   * @param logSegmentToCopy the {@link LogSegment} to copy from.
+   * @param indexSegmentToCopy the {@link IndexSegment} that contains the entries that need to be copied.
+   * @param duplicateSearchSpan the {@link FileSpan} in which to search for duplicates.
+   * @param startOffsetOfLastIndexSegmentForDelete  the start {@link Offset} of the index segment up until which delete
+   *                                                records are considered applicable.
+   * @return {@code true} if all the records in the index segment were copied. {@code false} if some records were not
+   * copied either because there was no more capacity or because a shutdown was initiated.
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there are any problems reading or writing to store components.
+   */
+  private boolean copyIndexSegmentData(LogSegment logSegmentToCopy, IndexSegment indexSegmentToCopy,
```

Comment:
can you rename this to copyLogSegmentDataPerIndexSegment(or copyDataPerIndexSegment). Current naming gives a notion that we copy just the index segment data and not the log data ;)